### ISSUE TITLE
NativeAOT metadata encodings

### ISSUE THREAD

**Issue by MichalStrehovsky** [dotnet/runtimelab#803]
NativeAOT metadata internally uses UTF-8 encoding to encode strings (things like names of types, or string literals, for reflection purposes).

The System.Runtime tests have a bunch of test cases with InlineDataAttribute that contains string literals with values that cannot be represented in UTF-8 (basically, invalid codepoints, such as "high surrogate" without a low surrogate).

When NativeAOT compiler generates metadata for these, we replace these with replacement characters. Unit test failures ensue.

Wondering whether we should roll our own UTF-8 encoder and decoder that simply passes the invalid codepoints through instead of crippling them. They can be represented - the encoders/decoders are just being told by the spec not to do that.

Alternatively we could also just ditch UTF-8 and store the UTF-16 code units as compressed integers. Or use compressed integers for the entire Unicode code point.

**Comment by jkotas** [comment]
Would `Encoding.EncoderFallback` and `Encoding.DecoderFallback` be capable of dealing with this?

**Comment by MichalStrehovsky** [comment]
I looked into that briefly, but EncoderFallbackBuffer operates on characters, not bytes, so I think it's going to hit the same problem. I'll look into it more.

**Comment by MichalStrehovsky** [comment]
https://github.com/dotnet/runtimelab/blob/70ba0a83896c9b278860aa31f2e3b4ab0fe642bd/src/libraries/System.Private.CoreLib/src/System/Text/EncoderFallback.cs#L251-L260

Nope, the fallback has to follow the same rules and whatever `char` it returns has to be encodable.

We'll have to make our own encoder unfortunately.

Decoder seems to have the same problem:

https://github.com/dotnet/runtimelab/blob/70ba0a83896c9b278860aa31f2e3b4ab0fe642bd/src/libraries/System.Private.CoreLib/src/System/Text/DecoderFallback.cs#L148-L150

So we also need a decoder.

**Comment by AntonLapounov** [comment]
[BlobWriter.WriteUTF8](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.blobwriter.writeutf8) and [BlobBuilder.WriteUTF8](https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.blobbuilder.writeutf8) allow writing unpaired surrogates and used for that purpose by Roslyn. For writing portable debug information they fall back to using UTF-16 in case of unpaired surrogates:
https://github.com/dotnet/roslyn/blob/1cca63b5d8ea170f8d8e88e1574aa3ebe354c23b/src/Compilers/Core/Portable/Serialization/ObjectWriter.cs#L500-L521

**Comment by MichalStrehovsky** [comment]
I should have kept this conversation on github, sorry.

Since most of the failing tests that looked like were caused by this (corrupt strings) were actually fixed by https://github.com/dotnet/runtime/pull/50088, we're left with a single failing test blocked on this.

We need a fast decoder for this. The built-in Utf-8 decoder is super fast and hand vectorized. It probably doesn't make sense to give up on those optimizations just for this corner case. It also doesn't make sense to replicate the optimizations and have to maintain a custom parser. I'm leaning towards a Won't fix.

```
From: Tomas Matousek  
Sent: Monday, March 29, 2021 18:43
To: Michal Strehovsky ; Chuck Stoner 
Cc: Andy Gocke 
Subject: RE: How does Roslyn encode UTF-8

Feel free to reopen the issue with additional info. Maybe we can persuade folks this time around.

From: Michal Strehovsky  
Sent: Monday, March 29, 2021 2:28 AM
To: Tomas Matousek ; Chuck Stoner 
Cc: Andy Gocke 
Subject: RE: How does Roslyn encode UTF-8

Thank you Chuck/Tomas!

The resolution of the “UTF8Encoding should support encoding/decoding of unpaired surrogates” issue fills me with sadness. I’ve also gone to the EncoderFallback/DecoderFallback dead end. It’s unfortunate because the Utf8 class has a ton of perf optimizations that are not worth replicating into a hand-rolled WTF8 encoder/decoder.

-Michal


From: Tomas Matousek  
Sent: Friday, March 26, 2021 17:44
To: Chuck Stoner ; Michal Strehovsky 
Cc: Andy Gocke 
Subject: RE: How does Roslyn encode UTF-8

MetadataReader supports round-tripping surrogates (aka WTF8 encoding). 

The reading behavior depends on the implementation of MetadataStringDecoder passed in: MetadataReader.cs (sourceroslyn.io). The default behavior is to treat unpaired surrogates as invalid input characters, but you can change that by implementing your own decoder.
The writer  writes WTF8 encoded string.

See also
UTF8Encoding should support encoding/decoding of unpaired surrogates · Issue #14785 · dotnet/runtime (github.com)

From: Chuck Stoner  
Sent: Friday, March 26, 2021 9:12 AM
To: Michal Strehovsky 
Cc: Andy Gocke ; Tomas Matousek 
Subject: RE: How does Roslyn encode UTF-8

Cc @Tomas


From: Chuck Stoner 
Sent: Friday, March 26, 2021 9:09 AM
To: Michal Strehovsky 
Cc: Andy Gocke 
Subject: RE: How does Roslyn encode UTF-8

Hey Michal.

It looks like Roslyn uses 
System.Reflection.Metadata.Ecma335.ScalarEncoder.Constant(object) for encoding and
System.Reflection.Metadata.BlobReader.ReadUTF8(int) for decoding

The decoding may not roundtrip correctly though – the value we get back from ReadUTF8() appears to be { 0xfffd, 0xfffd, 0xfffd }.



From: Michal Strehovsky  
Sent: Friday, March 26, 2021 3:54 AM
To: Chuck Stoner 
Cc: Andy Gocke 
Subject: How does Roslyn encode UTF-8

Hey Chuck,

Andy said you might know about this. I have a similar problem to what Roslyn needs to do in situations like:

class Program
{
    [Obsolete("\uD800")]
    static void Main()
    {
    }
}

The string literal above has a one half of a UTF-16 surrogate character in it. If I were to pass this to “Encoding.UTF8.GetBytes("\uD800")”, I would get a sequence of “EF BF BD” which when decoded back to a string maps to “\uFFFD” (Unicode replacement character).

However, when I open this program in ILDasm, I see:

.method private hidebysig static void  Main() cil managed
{
  .entrypoint
  .custom instance void [System.Runtime]System.ObsoleteAttribute::.ctor(string) = ( 01 00 03 ED A0 80 00 00 )

The sequence of UTF-8 bytes is “ED A0 80”, which is D800 encoded as (illegal) UTF-8, which can roundrip through a (non-spec-compliant) UTF-8 decoder back to \uD800.

I have a need for such non-spec compliant UTF-8 encoder and decoder. Do you know where to find Roslyn’s that I could steal?

Thanks,
-Michal
```

**Comment by AntonLapounov** [comment]
I think falling back to UTF-16 encoding as Roslyn does would meet described requirements in case we want to support this in future.  We could add a marker bit or use an invalid first byte (e.g., 0xff) as a marker that the value is encoded in UTF-16.
